src/minimega/vms.go

```diff
@@ -341,9 +341,34 @@ func (vms *VMs) Kill(target string) error {
 	})
 }
 
-// Flush deletes VMs that are in the QUIT or ERROR state, disconnecting them
+// Flush VMs matching target.
+func (vms *VMs) Flush(target string, cc *ron.Server) error {
+	return vms.Apply(target, func(vm VM, _ bool) (bool, error) {
+		if vm.GetState()&(VM_QUIT|VM_ERROR) == 0 {
+			return false, nil
+		}
+
+		log.Info("deleting VM: %v", vm.GetID())
+
+		if err := vm.Disconnect(cc); err != nil {
+			log.Error("unable to disconnect to cc for vm %v: %v", vm.GetID(), err)
+			return false, err
+		}
+
+		if err := vm.Flush(); err != nil {
+			log.Error("clogged vm %v: %v", vm.GetID(), err)
+			return false, err
+		}
+
+		delete(vms.m, vm.GetID())
```

Comment:
This needs to be synchronized otherwise it will panic.

Reply:
@glattercj to elaborate on this more, `vms.Apply` calls the provided function once per matching target VM found, each in their own Goroutine. As such, similar to how things are done in the (what's now called) `FlushAll` function below, the call to `delete` VMs from the map needs to be synchronized with a mutex lock.

So, for example:

```
func (vms *VMs) Flush(target string, cc *ron.Server) error {
  var mapLock sync.Mutex

  return vms.Apply(target, func(vm VM, _ bool) (bool, error) {
    ...

    mapLock.Lock()
    defer mapLock.Unlock()

    delete(vms.m, vm.GetID())

    return true, nil
  }
}
```

Reply:
Fixed in latest commit